hs-bindgen-runtime/src/HsBindgen/Runtime/FunPtr/Common.hs

```diff
@@ -0,0 +1,172 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- | Function pointer utilities and type class for converting Haskell functions
+-- to C function pointers.
+--
+-- This module provides a type class 'ToFunPtr' that allows for a uniform
+-- interface to convert Haskell functions to C function pointers.
+--
+-- This module also provides TH splices that generate FFI wrappers and
+-- 'ToFunPtr' instances that are called in different modules to paralellize
+-- compilation and compile time code generation.
+--
+module HsBindgen.Runtime.FunPtr.Common (
+    -- * Type class
+    ToFunPtr(..)
+  , FromFunPtr(..)
+
+    -- * Template Haskell FFI wrapper generation and 'ToFunPtr' instance
+    -- generation
+
+    -- ** Types
+  , allPrimTypes
+  , allPtrTypes
+  , allIOTypes
+    -- *** Common Types
+  , commonPrimTypes
+  , commonPtrTypes
+  , commonReturnTypes
+    -- ** Generate the instances
+  , generateInstance
+  ) where
+
+import Data.Void (Void)
+import Foreign qualified as F
+import Foreign.C.Types
+import GHC.Ptr qualified as Ptr
+import Language.Haskell.TH
+
+-- | Type class for converting Haskell functions to C function pointers.
+--
+class ToFunPtr a where
+  -- | Convert a Haskell function to a C function pointer.
+  --
+  -- The caller is responsible for freeing the function pointer using
+  -- 'F.freeHaskellFunPtr' when it is no longer needed.
+  --
+  toFunPtr :: a -> IO (F.FunPtr a)
```

Comment:
We could probably add a safe version of `toFunPtr` that takes care of the freeing:
 ```hs
 withToFunPtr :: a -> (F.FunPtr a -> IO b) -> IO b
 withToFunPtr x = bracket toFunPtr freeHaskellFunPtr
 ```